### PR TITLE
Autostyle fixes and new features

### LINKED ISSUE #8468: Auto-styling on widgets shoulnd't be available when having a point dataset aggregated by hexabins, squares or admin regions.

### LINKED ISSUE #8551: Improve auto-styling behaviour within the builder

### LINKED ISSUE #9872: Autostyle color picker on widgets options
@saleiva commented on [Thu Sep 01 2016](https://github.com/CartoDB/design/issues/176)

We had this originally, we removed, but is a common request now...

---

@arianaescobar commented on [Wed Sep 14 2016](https://github.com/CartoDB/design/issues/176#issuecomment-247079486)

Here's the UI proposal for this: https://invis.io/7D8LN1VEQ @urbanphes @saleiva 

The idea is to allow the users to specify which particular colors or color ramp they want to see when applying the widget autostyle from that particular widget settings.  

---

@urbanphes commented on [Thu Sep 15 2016](https://github.com/CartoDB/design/issues/176#issuecomment-247275741)

I think is good approach. 

---

@saleiva commented on [Mon Sep 26 2016](https://github.com/CartoDB/design/issues/176#issuecomment-249532460)

Love it, close and promote, please!
# SMOKE_WD16

### LINKED ISSUE #10118: Auto style on category widget doesnt get the legend refreshed
STR
- Get a legend for a choropleth map by styling a layer by value (number)
- Add a widget with an string column
- Click on the auto-style option. the legend should become a category one.
- When deactivating the auto-style the legend should be the numeric one.

I think we already talked about this, LMK if its duplicated.

### LINKED ISSUE #10408: Autostyling do not dissapear when removing widget
### STR
- Add a widget.
- Apply autostyling.
- Remove widget.
### Expected results

Style created with autostyling should have disappear once the widget is removed.
# SMOKE_WD27
# SMOKE_WD28
# SMOKE_WD29

### PR DESCRIPTION
**Pull request by xavijam**
Having almost all the functionality working, and pending only the necessary and needed tests. So, in short:
- [x] Represent auto-style in the Builder side (closes #8551).
- [x] Auto style on category widget doesn't get the legend refreshed (closes #10118).
- [x] Autostyle color picker on widgets options (closes #9872).
- [x] Expand widget definition with new style things.
  - Including backend changes (closes https://github.com/CartoDB/cartodb/issues/10252).
- [x] Remove AutoStyle button when an aggregation style is applied (closes #8468).
  - AutoStyle check should be disabled in that case.
- [x] Change widget type should "take" previous auto_style and widget_style properties.
- [x] ~~When widget column has changed, custom AutoStyle should be reseted if it was enabled~~.
- [x] Check if Time-series could manage auto-style: applying only proper color.
- [x] Add new options in the widget model (Deep-insights).
  - [x] Delete widget (closes https://github.com/CartoDB/deep-insights.js/issues/452).
  - [x] Edit widget (closes https://github.com/CartoDB/cartodb/issues/8003).
- [x] Hide quantification and bins from the fill component.
- [x] Disable auto-style if widget is removed (closes #10408).
- [x] ~~We should replace the use of `canFetch` (something new) to `shouldFetch` (something already working) in querySchemaModel and queryGeometryModel.~~
- [x] Having a non valid style type, disabled custom ramp component within widget appears weird. (https://github.com/CartoDB/cartodb/pull/10420/commits/00cd61fb6edd20f3d95c6679cb86c9b745bb32be)
- [x] Remove filter from widget if it is removed. (https://github.com/CartoDB/cartodb.js/pull/1521)
- [x] Review category fill component and its behaviour (category widget edition). (https://github.com/CartoDB/cartodb/pull/10420/commits/63ae3485096065890364e5fc5687d44ca39ae92a)
- [x] Review ramp fill component and its behaviour (histogram widget edition). (https://github.com/CartoDB/cartodb/pull/10420/commits/63ae3485096065890364e5fc5687d44ca39ae92a)
- [x] Talk with @donflopez about time-series bar color (green?).
- [x] Add all necessary tests.

But it is needed to have this [PR](https://github.com/CartoDB/deep-insights.js/pull/477) merged in DI before merging this one. And of course, it closes https://github.com/CartoDB/cartodb/issues/10255.

An acceptance guidelines is included [here](https://github.com/CartoDB/cartodb/pull/10420#issuecomment-267519787).

cc @CartoDB/frontend

### CODE REVIEW

File: lib/assets/javascripts/cartodb3/editor/style/style-converter.js [hunk 1]
```diff
@@ -41,6 +41,12 @@ function makeColorRamp (props, isTorqueCategory) {
       c.push('(' + _.map(props.domain, function (r, i) {
         return i + 1;
       }).join(', ') + ')');
+    } else if (props.static) {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/style/style-converter.js]: this is "hack"

File: lib/assets/javascripts/cartodb3/editor/widgets/widgets-form/schema/widgets-form-category-schema-model.js [hunk 2]
```diff
@@ -90,6 +90,22 @@ module.exports = WidgetsFormBaseSchema.extend({
         }
       }
     });
+
+    this._addAllStyleSchemaAttributes();
+
+    if (this.get('aggregation') !== 'count') {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/widgets/widgets-form/schema/widgets-form-category-schema-model.js]: Do we need this?

File: lib/assets/javascripts/cartodb3/editor/widgets/widgets-service.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
+var removeWidgetConfirmationTemplate = require('./delete-widget-confirmation.tpl');
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb3/editor/widgets/widgets-service.js]: Let's review if you find anything wrong or a better approach

**Comment by nobuti** [comment]
Ready for code review.

**Comment by nobuti** [comment]
## What should be tested

Read the PR description in order to get the complete idea, however here you have some points that they deserve your attention:

- [ ] Check that once a category or histogram widget, the auto style button is there and working.
- [ ] Once auto style is applied, the map is refreshed with the styles inferred from the widget.
- [ ] If auto style is applied and you change some layer's style, the auto style disables. All widgets reflect this state (no auto styling).
- [ ] The case above also applies when you change the style's aggregation. Only points has auto style allowed. 
- [ ] If you disable the auto style, the map get back to the last state, style speaking.
- [ ] Auto style disable legends. Autostyling a category widget disables category legends. The same thing happens for histogram and choropleths.
- [ ] You can customise the widget's bar. Once you are in the widget's definition form, a fill component handles the bars' color (category, histogram and time-series).
- [ ] If you applied the auto style on a widget, you can customise the ramp. In the widget's definition form, an enabler form control hides the fill component that handles it.
- [ ] If a widget has a filter and it's removed, the map refreshes to show the full data range from the deleted widget. That is, the filter is gone and the map refreshes.

The auto style feature is behind a feature flag, and we plan to ship it that way initially. Because of this, more than likely all non critical issues that raise from the acceptance will be fixed in other issues. **So it's extremely important to test everything is working right for users with no auto style flag**. 

Please, don't hesitate to ping me if you have any questions.

**Comment by nobuti** [comment]
It's in ded08.

**Comment by saleiva** [comment]
Finding some issues that I'd put under the milestone
https://github.com/CartoDB/cartodb/issues/11084

**Comment by saleiva** [comment]
Ok, tested. Looks to be working pretty well. Just a few things added to the milestone

**Comment by nobuti** [comment]
I'm uploading this to ded04, ded08 seems to be broken again.

**Comment by javitonino** [comment]
Testing and everything works well. Just a couple of issues/questions:
 - Shouldn't the histogram widget reflect that auto-style is applied by changing the color of its bars to reflect the color ramp applied? Right now, it keeps the default color for all bars.
 - In the embed, legends are not disabled when applying autostyle. In the builder, they disappear (as expected). Should the legends disappear in the embed as well?

I have also tested everything I could think of with the FF off, and nothing breaks. And even if the things I mention here are bugs, I think we could deploy this and open new tickets for those. But, up to you, leaving this in acceptance meanwhile.

**Comment by nobuti** [comment]
@javitonino about the first issue: https://github.com/CartoDB/cartodb/issues/11095. I'll take a look about the second, it's going to be tricky, because in the embed we haven't got the needed connections between legends, style and widgets 😞

**Comment by javitonino** [comment]
Split issue to https://github.com/CartoDB/cartodb/issues/11098

So this PR is ready for merging after conflicts are resolved ✨

**Comment by nobuti** [comment]
@matallo what do you think about https://github.com/CartoDB/cartodb/pull/10420/commits/dcb30498dfd7db818197b4a3ed9d54b2ac5ffcab? This solves all my troubles, I had to run this task in order to copy all needed files, that's why the bundle was empty, all files were moved when merging. Not sure if we should add this task as part of the dev task.

**Comment by nobuti** [comment]
👍  on demand then, are you ok with the `pre_client` task augmentation to include those sub tasks or should I revert the change?

**Comment by matallo** [comment]
I'm ok with that as long as we set a reminder to remove those in the near future, in any case if we registered a separate task we should remove that, too

**Comment by nobuti** [comment]
🚀

**Pull request merged**